test/src/test/kotlin/graphql/nadel/tests/next/NadelIntegrationTest.kt

```diff
@@ -0,0 +1,317 @@
+                    .makeExecutableSchema(
+                        SchemaParser().parse(service.underlyingSchema),
+                        RuntimeWiring.newRuntimeWiring()
+                            .also(service.runtimeWiring)
+                            .build(),
+                    ),
+            )
+            .instrumentation(
+                object : Instrumentation {
+                    override fun instrumentExecutionResult(
+                        executionResult: ExecutionResult,
+                        parameters: InstrumentationExecutionParameters,
+                        state: InstrumentationState?,
+                    ): CompletableFuture<ExecutionResult> {
+                        return CompletableFuture.completedFuture(
+                            executionCapture.capture(
+                                service = service.name,
+                                query = parameters.query,
+                                variables = parameters.variables,
+                                result = executionResult,
+                            ),
+                        )
+                    }
+                },
+            )
+    }
+
+    open fun makeInstrumentation(): NadelInstrumentation {
+        return object : NadelInstrumentation {
+        }
+    }
+
+    open fun getTestData(): TestData {
+        return Class.forName(this::class.qualifiedName + "Data")
+            .getDeclaredConstructor()
+            .newInstance() as TestData
+    }
+
+    open fun assert(result: ExecutionResult) {
+    }
+
+    private fun assertServiceCalls(testData: TestData) {
+        // Unmatched calls, by the end of the function both should be empty if they're matched
+        val unmatchedExpectedCalls = testData.calls.toMutableList()
+        val unmatchedActualCalls = executionCapture.calls.toMutableList()
+
+        unmatchedActualCalls.forEachElementInIterator { iterator, actualCall ->
+            fun getCanonicalQuery(query: String): String {
+                return AstPrinter.printAstCompact(Parser().parseDocument(query))
+            }
+
+            val actualQuery = getCanonicalQuery(actualCall.query)
+            val actualVariables = actualCall.variables
+
+            val matchingCalls = unmatchedExpectedCalls
+                .filter { expected ->
+                    actualCall.service == expected.service
+                        && getCanonicalQuery(expected.query) == actualQuery
+                        && jsonObjectMapper.readValue<JsonMap>(expected.variables) == actualVariables
+                }
+
+            // Multiple matches is ok, we match one at a time though
+            if (matchingCalls.isNotEmpty()) {
+                unmatchedExpectedCalls.remove(matchingCalls.first())
+                iterator.remove()
+            }
+
+            // todo: match delayed responses here too
+        }
+
+        assertTrue(unmatchedExpectedCalls.isEmpty() && unmatchedActualCalls.isEmpty())
```

Comment:
will this assert fail if the number of expected call in the test data is different than the actual number of calls at execution time? Perhaps add a description to this assertion to make the reason of the failure more clear.

Reply:
Yeah I'll update it to be clearer, but yes it will fail if call numbers are different, or are different content wise.